### ISSUE TITLE
[Issue]:Not useing correct music cover when there is mutiple covers used in an album

### ISSUE THREAD

**Issue by MisterGenius0** [jellyfin/jellyfin#8570]
### Please describe your bug

When I Add an album with multiple covers used in different songs, jellyfin picks a random cover in the song metadata and makes all songs that cover artwork. 

for example, I have 6 disks in an album, and each disk has a different cover used, in the metadata, they are all correct with 6 other covers associated with the 6 disks. jellyfin will only pick 1 cover and assign all songs to have that selected cover.

I have tried to use different metadata fetchers MusicBrainz and TheAudioDB but the issue still persists.

I have noticed that if you give the album a cover then jellyfin will assign all songs that the album artwork ignoring the actual metadata and using the wrong image when you play the song.

file structure
AllMusic
---Half-Life Alyx Soundtrack (Artist)
        ---Half-Life Alyx Soundtrack (Album)
                --Disk1
                --Disk2
                ...
                --....Disk11the 
                
                cover.jpeg

### Jellyfin Version

10.8.0

### Environment

```markdown
- OS: Windows
- Virtualization:
- Clients:browser, UWP app
- Browser: Firefox
- FFmpeg Version:
- Playback Method:
- Hardware Acceleration:
- Plugins: none
- Reverse Proxy: none
- Base URL: none
- Networking:
- Storage: Local
```

### Please attach any screenshots here

![Metadata cover image](https://user-images.githubusercontent.com/39445650/196427339-02d6bfa5-f105-4e8d-8a71-35382ce0b6ca.png)

Here you can see it's playing the song "Only One" in the Half-life Alyx soundtrack.
I'm showing that windows' default music player is doing it correctly (Yellow-green image), showing the correct image in the metadata. 
I'm also using mp3 tag to display the actual metadata for this song proving that jellyfin has the wrong metadata for this song (Blueish image that's for disk 10, not disk 2)

This album has no album artwork I set up it and found the metadata using the Image Extractor, the other methods like MuisicBrainz are turned off for this example

![Screenshot 2022-10-18 053803](https://user-images.githubusercontent.com/39445650/196431538-d8d38ff4-8091-45a7-9a49-355eb5c9b309.png)

when I added a cover artwork in the files and refreshed the metadata the currently playing song metadata was still wrong

![Screenshot 2022-10-18 054230](https://user-images.githubusercontent.com/39445650/196432418-537eb5d9-e91b-44d8-a766-70ab84206202.png)
File structure of my songs

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kyz** [comment]
I can confirm this bug (or perhaps misfeature?)

Expected Behaviour:
1. in a music album, if a song has embedded cover art, that should be the song's primary image
2. otherwise, it should be the album's cover art ("cover.jpg", "folder.jpg", etc. in the album's directory)

Actual Behaviour:
1. when [`AudioImageProvider.GetImage()`](https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Providers/MediaInfo/AudioImageProvider.cs#L77) runs, it creates a cached image based on album name and premier album artist
2. all other songs in the album, with embedded cover art, hash to this same image and reuse it
3. only songs with _no_ embedded cover art use the album's cover art ("cover.jpg", "folder.jpg", etc.)

I understand this is done as some kind of optimisation, as most albums will have all songs with the same cover art, and Jellyfin neither wants to store this, nor even bother scanning it. However, this prevents a number of use cases, such as having a multi-disc album where the songs of each disc have different cover art, or having unique art per song.

What could be done instead:
1. Extract **every** audio file's embedded art
2. Create the per-album cached file as before (and include disc number into the cache filename)
3. If the cached file exists **and** the current audio file's embedded art matches the cached file, use the cached file, **otherwise** save and use the unique art.

Here's a comparison of how Jellyfin, VLC and Windows Explorer handles embedded art:

![jellyfin-cover-choice](https://user-images.githubusercontent.com/530810/196806043-e1057b11-baa7-4938-9b19-6a074bd63a52.png)

* Windows Explorer always shows embedded art, or a default icon if the audio file does not have embedded art (also it does not support .ogg embedded art)
* VLC does the same, but shows cover.jpg from the enclosing folder if the audio file does not have embedded art
* Jellyfin is as described above. Whichever audio file in the album it examines first, if it contains cover art, that sets the cache-file for anything with the same albumname-albumartist1, except if the audio file does not have embedded art in which case it will use cover.jpg from the enclosing folder

**Comment by Happyrobot33** [comment]
I agree that this is also an issue I am experiencing. One time I was able to get some of the audio files to have correctly set images, but now I cant get them to be unique at all

**Comment by Happyrobot33** [comment]
This actually seems already somewhat supported, but impropely determined, by these lines in the else statement
[// If it's an audio book or audio podcast, allow unique image per item
filename = item.Id.ToString("N", CultureInfo.InvariantCulture) + ".jpg";](https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Providers/MediaInfo/AudioImageProvider.cs#L132)

**Comment by Happyrobot33** [comment]
This is still an issue in the latest release

**Comment by AkelaDev** [comment]
I can also attest to this. I have several single songs with embedded album art. If the singles have the same artist, then even though they have different embedded art, all of them will adopt the embedded art from on of the songs. This seems to sometimes happen even without having an artist in common. I'd like if Jellyfin could just read directly from the embedded art for each song individually.

**Comment by gummykage** [comment]
This used to work... Why did it stop working? Or maybe it randomly worked.

**Comment by gummykage** [comment]
![image](https://github.com/jellyfin/jellyfin/assets/9695713/b151b83a-eefa-407d-b31a-ea408125e61f)
This not actually a playlist, I just like naming a collection of songs I want into the same album name with the word playlist after it. For whatever reason here it extracted the images from each song. I changed the tag for album artist to various after it extracted the covers for each song so it would display the artist under each track rather than a huge ugly clump the goes over parts of the tracks you can't read it.
![image](https://github.com/jellyfin/jellyfin/assets/9695713/f5de1819-1f05-473f-8d47-f86ec103581c)

**Comment by gummykage** [comment]
But after changing the album artist name to one artist (various in this case) another jellyfin instance takes the image from the first track and hashes it over all the other tracks. Please someone consider fixing this to extract images from each track or use the coverart when none is available. Would solve so many issues with people with hundreds of compilation cds like the Top 100 ones.

**Comment by Happyrobot33** [comment]
Bumping as still seeming to be an issue

**Comment by yamanobe96** [comment]
Sorry, this problem still exists

In recent years, soundtrack sales have become mainstream, and in recent years there has been an increase in the number of cases in which each song in an album has different cover art.
I think this is a problem that needs to be solved

**Comment by NicholasFlamy** [comment]
Bump because WTF, come on.

**Comment by felix920506** [comment]
I just tested on master and this doesn't seem to be a problem. If an external cover file is provided then it is used. If not then it uses the cover of the first track. There doesn't seem to be any randomness in this.

I tested with a test folder consisting of 376 individual tracks and a cover.png file, with all 377 images being different.

When viewing in the player, each song also retains its own cover art.

**Comment by NicholasFlamy** [comment]
1. 

> I just tested on master and this doesn't seem to be a problem. If an external cover file is provided then it is used. If not then it uses the cover of the first track. There doesn't seem to be any randomness in this.
> 
> I tested with a test folder consisting of 376 individual tracks and a cover.png file, with all 377 images being different.
> 
> When viewing in the player, each song also retains its own cover art.

So it was using the same cover image no matter what on my music library even when disabling metadata providers, deleting the image with the metadata manager, refresh metadata (all 3 options including the replace all with images checked.) So now I deleted that one and created a new one with all the metadata providers disabled while creating. This one doesn't show any cover image at all. Every track has a jpg file encoded with mozjpeg at 80% quality at a resolution of 1200x1200 (so the files are from 50 KB to 300 KB) embedded into the file using Picard.

**Comment by NicholasFlamy** [comment]
> This one doesn't show any cover image at all

After some combination of refreshing metadata and search for missing now it has cover images, even with all the metadata providers disabled and it's not pulling from the embedded data.

**Comment by solidsnake1298** [comment]
Did you also disable the embedded image extractor providers?

**Comment by NicholasFlamy** [comment]
> Did you also disable the embedded image extractor providers?

I had it disabled on creation (99% sure) I just now enabled it (I think) and I think that might have caused it to pull metadata from the web.

**Comment by NicholasFlamy** [comment]
This was purely caused by enabling this and search for missing metadata.
![Screenshot_20240206_231949_Jellyfin](https://github.com/jellyfin/jellyfin/assets/30300649/c1380771-995e-4ebe-81b7-fc234a4655ee)

**Comment by felix920506** [comment]
@NicholasFlamy If you disable the Image Extractor then not having covers is going to be expected behavior.

**Comment by NicholasFlamy** [comment]
> @NicholasFlamy If you disable the Image Extractor then not having covers is going to be expected behavior.

When I enabled it, it pulled data from the internet (without changing anything else.)

**Comment by AkelaDev** [comment]
> I just tested on master and this doesn't seem to be a problem. If an external cover file is provided then it is used. If not then it uses the cover of the first track. There doesn't seem to be any randomness in this.
> 
> I tested with a test folder consisting of 376 individual tracks and a cover.png file, with all 377 images being different.
> 
> When viewing in the player, each song also retains its own cover art.

I am currently testing this right now. I know I have thought I had fixed the bug before only for it to pop up again later, so I'm trying to find a consistent way to replicate.

I made an new folder on my NAS and copied several songs from several different albums and soundtracks to it. Each had their own embedded album cover.  Made a library in Jellyfin just of these, and only activated Media Extractor. All the songs indeed had their own image.

Did the same thing but used MP3 tag to change all of their Album tags to the same thing. Made a new library from just this folder. All the songs had their own image again.

I had a hunch and changed the Album Artist tag to the same thing and refreshed metadata of the library. _Then_ all songs shared the same cover art despite each having their own embedded album art.

So I have a hunch that it has to do with songs having the same Album Artist tag even if they're in a different album. Since I was able to replicate it by refreshing metadata I'm trying to see if it also does it when first creating a library.

**Comment by AkelaDev** [comment]
Okay I can confirm now. Tested with a handful of songs in a fresh library. If two or more songs have the same Album and Album Artist tag, and at some point the user scans the library with "Refresh all metadata" chosen and "Replace existing images" checked, then those songs will share the same album image in the Jellyfin user interface. This is despite having a different embedded image.

By making the Album and Album Artist tags the same in the file tags, then doing the above scanning settings, I was able to reliably replicate the issue. Below is an image of my Jellyfin library, with only the ones I changed have the same tags sharing the same image.

![Screenshot_20240207-015757](https://github.com/jellyfin/jellyfin/assets/99080076/f6bdb473-fa95-4878-9975-e56690824da3)
![Screenshot_20240207-015728](https://github.com/jellyfin/jellyfin/assets/99080076/227f0e54-f454-4bf4-97cb-a7372bde7123)

Now because this is so specific maybe it's not worth pursuing. Personally I think it's reasonable for a user to expect a scan, no matter the specific settings, to replace embedded album art. If an album has unique art on each song, then this bug would replace the unique art with a common one.

**Comment by Happyrobot33** [comment]
I definetly think this should be fixed still, or even at the very least an option to respect the unique art. A major reason why i run into this issue is the use case of podcasts, where each "season" might have different season art on the file, but the artist always stays the same since its the same hosts

**Comment by felix920506** [comment]
This is no longer an issue in Master. 
According to my testing:
- Each song retains its own cover when they are displayed 
- The album can have a different cover from all the songs if provided by the user
- The cover art of the first track is grabbed as a stand in if nothing is provided by the user

**Comment by felix920506** [comment]
Closing this issue as it is already fixed in master. 
Anyone is welcome to open a new report if this behavior reappears in the future **IN MASTER**
further reports of this issue on 10.8.x will be closed.

**Comment by NicholasFlamy** [comment]
But then why does Jellyfin download the album cover from the internet even if all internet metadata providers are disabled and I only enable "Image Extractor" on 10.8.13? It shouldn't download from the internet if that is disabled. I also deleted the music libraries and ran the clean cache scheduled task. I'll try and chmod 777 the jellyfin config, cache etc. directories to see if it's a permission error. I'll also try and nuke the cache directory. But before I nuke that maybe y'all can tell me the risks?

**Comment by NicholasFlamy** [comment]
I know it's not the embedded art because the embedded art is for the Deluxe version and it's slightly different from the normal but it doesn't have Deluxe in the album name so Jellyfin downloaded the standard art. All the songs in this album were tagged with Picard and have matching images. Also, I only have 2 albums in this music library (this is for testing purposes) by 2 different artists and both have this issue.

**Comment by NicholasFlamy** [comment]
I'll attempt to build the master branch and test it if nuking the cache doesn't work to see if this issue is also fixed but I'll wait on that.

**Comment by NicholasFlamy** [comment]
When I run the clean cache scheduled task, the incorrect album cover can still be found in cache (I poked around and can still see the files.)

**Comment by NicholasFlamy** [comment]
Okay so Jellyfin has a problem with getting rid of old cache. After nuking cache and artist metadata and chmod 777 just in case now it works as intended. 

`rm -r /mnt/apps/data/jellyfin/cache/*`
`rm -r /mnt/apps/data/jellyfin/config/metadata/artists`
`chmod -R 777 /mnt/apps/data/jellyfin/`

FYI I'm on TrueNAS Scale so that's why those directories are under /mnt. Also I did this from a root shell but I could have done sudo for those commands.

**Comment by AkelaDev** [comment]
> Closing this issue as it is already fixed in master. Anyone is welcome to open a new report if this behavior reappears in the future **IN MASTER** further reports of this issue on 10.8.x will be closed.

Could you explain a little of what that means. What version is master on? Does that mean we should soon see a full download on the site where this is already fixed?

**Comment by solidsnake1298** [comment]
> > Closing this issue as it is already fixed in master. Anyone is welcome to open a new report if this behavior reappears in the future **IN MASTER** further reports of this issue on 10.8.x will be closed.
> 
> Could you explain a little of what that means. What version is master on? Does that mean we should soon see a full download on the site where this is already fixed?

Master is the future 10.9 release.

**Comment by kyz** [comment]
@felix920506 can you explain how it's fixed in master, if [the problem code](https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Providers/MediaInfo/AudioImageProvider.cs#L111-L139
) has remained the same in master for years? I don't see any fix applied.

The code still appears to use a cache for images, and still uses a key of album name + premier artist name... for any audio that is not podcast/audiobook.

The testing above showed an album with different images per track... but also EACH TRACK WAS A DIFFERENT ARTIST.... which mean that the bug isn't shown.

Has anyone tested where each track is in the SAME album and has the SAME artist, but a DIFFERENT image, and confirmed Jellyfin shows different images for each track?

**Comment by felix920506** [comment]
@kyz well I tested it on master and it didn't happen

**Comment by felix920506** [comment]
To anyone who is unhappy with the decision of closing this issue:
1. This will not be reopened. Please open a new issue with concrete reproduction steps if you manage to reproduce it.
2. We are volunteers. No one is obligated to service you and/or fix your specific problem.
3. Arguing will get you nowhere.

**Comment by AkelaDev** [comment]
> To anyone who is unhappy with the decision of closing this issue:
> 
>     1. This will not be reopened. Please open a new issue with concrete reproduction steps if you manage to reproduce it on master.
> 
>     2. We are volunteers. No one is obligated to service you and/or fix your specific problem.
> 
>     3. Arguing will get you nowhere.

No worries. I can pull master and web server but don't understand how to set it up so I'll wait until it's on the next 10.x release to confirm for myself. Thanks for looking into it.

**Comment by NicholasFlamy** [comment]
> > To anyone who is unhappy with the decision of closing this issue:
> > ```
> > 1. This will not be reopened. Please open a new issue with concrete reproduction steps if you manage to reproduce it on master.
> > 
> > 2. We are volunteers. No one is obligated to service you and/or fix your specific problem.
> > 
> > 3. Arguing will get you nowhere.
> > ```
> 
> No worries. I can pull master and web server but don't understand how to set it up so I'll wait until it's on the next 10.x release to confirm for myself. Thanks for looking into it.

![Screenshot 2024-02-07 223431](https://github.com/jellyfin/jellyfin/assets/30300649/1ee910d6-c340-4aa1-962c-58e3861df358)
![Screenshot 2024-02-07 223457](https://github.com/jellyfin/jellyfin/assets/30300649/b3ca6310-253f-4f7e-a9fb-f05b02333a8a)


P.S. sorry for responding to this closed issue. I just wanted to help this guy out.

**Comment by yamanobe96** [comment]
I read that this problem has been solved, so please tell me how to set it up.
I get mp3 files from youtube and embed the video thumbnail into the artwork to create a collection.
All songs have different artwork embedded.
I tried registering my library in jellyfin, but the artwork was overwritten by the artwork of the first song and the artwork of the other songs disappeared.
![001](https://github.com/user-attachments/assets/6117b181-b378-4b9e-a779-e3e90c0835ea)
![002](https://github.com/user-attachments/assets/bae18a9b-d9ec-43d5-b37d-46c4a41cb2f8)
![003](https://github.com/user-attachments/assets/d65dc671-d9d7-422f-a1c4-6923cd10b018)
![004](https://github.com/user-attachments/assets/7590785a-3569-447c-aac1-57d1e05dfd68)

**Comment by NicholasFlamy** [comment]
> I read that this problem has been solved, so please tell me how to set it up. I get mp3 files from youtube and embed the video thumbnail into the artwork to create a collection. All songs have different artwork embedded. I tried registering my library in jellyfin, but the artwork was overwritten by the artwork of the first song and the artwork of the other songs disappeared. img img img img

If this server was set up on an older version then it probably won't update the assets unless rescanned. It might also require clearing Jellyfin's cache.

First, try rescanning with replace all metadata:
![image](https://github.com/user-attachments/assets/3749a930-daff-413b-9e0e-472edd69991d)

If that doesn't do the trick then could try clearing the cache and then rescanning:

## Warning: I'm not associated with Jellyfin and clearing the cache by using these commands can break something if you mess up, don't run them if you don't understand them:
https://github.com/jellyfin/jellyfin/issues/8570#issuecomment-1933088077
(P.S. I shouldn't have to tell you, but replace the path in the commands with your own, if you don't know what that is then do not mess around with this.

**Comment by yamanobe96** [comment]
https://github.com/jellyfin/jellyfin/issues/8570#issuecomment-2436397875

Thank you for letting me know, but even after clearing the cache and recreating all the databases, the artwork remains fixed for each album.

**Comment by NicholasFlamy** [comment]
> [#8570 (comment)](https://github.com/jellyfin/jellyfin/issues/8570#issuecomment-2436397875)
> 
> Thank you for letting me know, but even after clearing the cache and recreating all the databases, the artwork remains fixed for each album.

I see that this issue is linked to PR #11227. That PRis not merged so maybe the functionality isn't in Master yet? That or, your media is unusual.
